Zend/zend_gc.c

```diff
@@ -705,6 +723,25 @@ static void gc_scan_black(zend_refcounted *ref, gc_stack *stack)
 			zval *table;
 			int len;
 
+			if (UNEXPECTED(GC_FLAGS(obj) & IS_OBJ_WEAKLY_REFERENCED)) {
+				zend_weakmap_get_object_values_gc(obj, &table, &len);
+				n = len;
+				zv = table;
+				for (; n != 0; n--) {
+					ZEND_ASSERT(Z_TYPE_P(zv) == IS_PTR);
+					zval *value = (zval*) Z_PTR_P(zv);
+					if (GC_FROM_WEAKMAP_KEY(value) && Z_OPT_REFCOUNTED_P(value)) {
```

Comment:
In all phases but the first one, we only follow the virtual reference if the entry is marked with GC_FROM_WEAKMAP_KEY